DOC: Show probability density functions as math

MartinThoma opened the pull request.

MartinThoma commented:
Is this kind of change wanted? If so, I would do it for the complete file for all pdfs.

MartinThoma commented:
I think the rendered formula is MUCH easier to read than the pure text.

rgommers commented:
The html is easier to read, but for the source code itself and in a terminal it becomes much harder. At the very least these expressions should be moved to comments in ``_pdf`` rather than removed, to not make maintenance work harder.

@ev-br opinion?

pv commented on scipy/stats/_continuous_distns.py:
```diff
@@ -2163,19 +2163,21 @@ def fit(self, data, *args, **kwds):
+    .. math::
 
-    for ``x >= 0``, ``a > 0``, and ``c != 0``.
+        f(x, a, c) = \frac{|c| \cdot x^{c \cdot a-1} \cdot e^{-x^c}}{\Gamma(a)}
```
Multiplication sign is always omitted in standard math notation.

pv commented on scipy/stats/_continuous_distns.py:
```diff
@@ -2224,20 +2227,21 @@ def _entropy(self, a, c):
 
-        genhalflogistic.pdf(x, c) =
-            2 * (1-c*x)**(1/c-1) / (1+(1-c*x)**(1/c))**2
+        f(x, c) = \frac{2 \cdot {(1-c \cdot x)}^{\frac{1}{c-1}}}{{\left (1+{(1-c \cdot x)}^{1/c} \right)}^2}
```
Use `[` and `]` instead of `\left(` and `\right)`

pv commented on scipy/stats/_continuous_distns.py:
```diff
@@ -2224,20 +2227,21 @@ def _entropy(self, a, c):
+        f(x, c) = \frac{2 \cdot {(1-c \cdot x)}^{\frac{1}{c-1}}}{{\left (1+{(1-c \cdot x)}^{1/c} \right)}^2}
 
-    for ``0 <= x <= 1/c``, and ``c > 0``.
+    for :math:`0 \leq x \leq 1/c`, and :math:`c > 0`.
```
It's usually just `\le`

pv commented on scipy/stats/_continuous_distns.py:
```diff
@@ -2163,19 +2163,21 @@ def fit(self, data, *args, **kwds):
+        f(x, a, c) = \frac{|c| \cdot x^{c \cdot a-1} \cdot e^{-x^c}}{\Gamma(a)}
 
-    `gengamma` takes ``a`` and ``c`` as shape parameters.
+    for :math:`x \geq 0`, :math:`a > 0`, and :math:`c \neq 0`.
```
Usually just `\ge`, `\ne`

pv commented on scipy/stats/_continuous_distns.py:
```diff
@@ -2224,20 +2227,21 @@ def _entropy(self, a, c):
 
-        genhalflogistic.pdf(x, c) =
-            2 * (1-c*x)**(1/c-1) / (1+(1-c*x)**(1/c))**2
+        f(x, c) = \frac{2 {(1-c x)}^{\frac{1}{c-1}}}{{\left (1+{(1-c x)}^{1/c} \right)}^2}
```
Better (APS) latex styling, and easier to read as text:
`\frac{2 (1 - c x)^{1/(c-1)}}{[1 + (1 - c x)^{1/c}]^2}`

MartinThoma replied:
I think mine looks much better in LaTeX. See

![Imgur](http://i.imgur.com/fnNAcVm.png)

the first one is mine, the second one is the LaTeX you suggest, the thirds is a mixture `\frac{2 {(1 - c x)}^{1/(c-1)}}{{\left[1 + {(1 - c x)}^{1/c} \right]}^2}`

I don't care about the type of brace. Using `\left` / `\right` makes ist easier to read. Enclosing the base of the exponent in curly braces allows LaTeX to recognize the term correctly (otherwise, LaTeX will only take the last character).

Anyway, I think all 3 options are much better than what we currently have.

pv replied:
\left,\right make it more difficult to read as *text*, which is also a problem with the extra frac in the exponent.

I'm pretty sure I prefer my version over the others, but OTOH I'm used to the APS journal style.

ev-br commented:
@rgommers How about (i) we do keep pseudocode in comments (maybe even add a python-readable string attribute?), (ii) keep the TeX notation minimalistic (no `\left(` and `\right)`, certainly no `cdot` or `\times` for multiplication), (iii) enforce some readability standards in TeX formulas, almost pep8-ish (no excessive `{`s, spaces around binary operators etc). 

`genhalflogistic` example is nice: while I'm generally used to reading/writing TeX maths, I find the original version unreadable as text, but Pauli's version, https://github.com/scipy/scipy/pull/7756#discussion_r134121393, is IMO very readable.

MartinThoma commented:
@ev-br @pv Ok, I've adjusted my pull request, added the rules to the commit message (so that we don't forget it later on) and went through some more examples. There is one other thing where a similar preference should be decided:

    exp(whatever)

vs

    e^{whatever}

person142 commented:
It would probably be good to document @ev-br's points (i)-(iii) somewhere since the discussion about LaTeX in docstrings keeps coming up.

MartinThoma commented:
@ev-br @pv: Any opinion on

1. `\exp(abc)` vs `e^{abc}`?
2. Put the pseudocode in a docstring in `_pdf` vs put it as a comment in `_pdf?
3. Traditional TeX vs mixed unicode: I could also put special character as unicode signs in the document - mathjax can manage that. So `f(x, a, c) = \frac{|c| x^{c a-1} e^{-x^c}}{\Gamma(a)}` would be `f(x, a, c) = \frac{|c| x^{c a-1} e^{-x^c}}{Γ(a)}` instead.

What do you think about this?

ev-br commented:
1. I would try following e.g. https://github.com/scipy/scipy/pull/7756#issuecomment-324145453 

2. Docstrings of _pdf methods are not very user visible, so I'm not sure it matters much.

3. No. If it's TeX, then it's \Gamma or \gamma. TOWDII and all that.

BTW, with gamma functions specifically, $\Gamma$ and $\gamma$ are different things. Please double-check.

MartinThoma commented:
@ev-br I went through all changes, looking for `\gamma` / `\Gamma`. I think I didn't change it. While I didn't check if it was correct before, at least I should not have introduced new errors.

I think I changed the usage of `\exp` / `e^{...}` according to the link you shared. I added the link to my commit message to document this decision.

From my side, it looks ready (I'll remove the "WIP" from the commit message in a second)

rgommers commented on scipy/stats/_continuous_distns.py:
```diff
@@ -5348,14 +5645,16 @@ class argus_gen(rv_continuous):
         """
         Return PDF of the argus function
+
+        argus.pdf(x, chi) = chi**3 / (sqrt(2*pi) * Psi(chi)) * x * sqrt(1-x**2) * exp(- 0.5 * chi**2 * (1 - x**2))
```
please keep lines under 80 characters

rgommers commented:
@MartinThoma there are a couple of test errors, both PEP8 and ``SyntaxError: invalid escape sequence \e``. For the latter you'll want to use a raw docstring I think (``r"""``)

rgommers commented:
I'm now okay with how this looks. @ev-br, @pv please have another look if you can. I'll add it to the 1.0 milestone.

MartinThoma commented:
@rgommers Could you please point me where those are? I only see PEP8 issues where I didn't edit anything.

For the non-ascii character(s) I also don't know where to look. I tried

    $ grep -P '[^\x00-\x7f]' _continuous_distns.py

but it doesn't show anything.

MartinThoma commented:
oops, found the PEP8 issue. I'll fix it in 1 minute.

pv commented:
See the error messages in the travis-CI logs (click the red X next to 
the commit or at bottom of the page).

MartinThoma commented:
@pv I don't see a Travis log. Only appveyor. And I can't access any details of appveyor, but only see

>  AppVeyor was unable to build non-merg...

pv commented:
If it doesn't show up here, you can go to https://travis-ci.org/scipy/scipy/pull_requests and check the latest one corresponding to yours.

MartinThoma commented:
@pv I only see https://travis-ci.org/scipy/scipy/builds/273560231 which is outdated and only shows (already fixed) pep8 issues.

pv commented:
Look at the python3.6 failures for the link you posted.

larsoner commented:
@MartinThoma can you rebase your changes? This will allow the CIs to run properly.

larsoner commented on scipy/stats/_continuous_distns.py:
```diff
@@ -5331,15 +5641,23 @@ class argus_gen(rv_continuous):
+    .. math::
+
+        f(x, \chi) = \frac{\chi^3}{\sqrt{2\pi} \Psi(\chi)} x \sqrt{1-x^2}
+                     \exp(- 0.5 \chi^2 (1 - x^2))
```
The old (outdated) error is here:

https://travis-ci.org/scipy/scipy/jobs/273560235#L1718

But the hint is still correct -- you have `\e` but no `r"""`

rgommers commented:
Rebased and fixed the syntax error

rgommers commented:
Okay, this is all green and looks good, so merging. Thanks @MartinThoma, all!

rgommers commented:
I went through every distribution in the html docs, there are a couple of inconsistencies left (mostly not due to this PR). Will send a follow-up PR.

@MartinThoma if you're interested in doing the same for the discrete distributions, that will be very welcome.

MartinThoma commented:
Sure,I can do so. Just not this weekend as I'm currently on a hiking trip

On 16 Sep 2017 5:12 AM, "Ralf Gommers" <notifications@github.com> wrote:

> I went through every distribution in the html docs, there are a couple of
> inconsistencies left (mostly not due to this PR). Will send a follow-up PR.
>
> @MartinThoma <https://github.com/martinthoma> if you're interested in
> doing the same for the discrete distributions, that will be very welcome.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scipy/scipy/pull/7756#issuecomment-329941367>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABlNBb-H2oSN79ya3aYVNsehg_aWSj26ks5sizyugaJpZM4O8Puu>
> .
>

rgommers commented:
Great! No hurry, whenever. Just nice for consistency.
